src/System.Collections/src/System/Collections/BitArray.cs

```diff
@@ -266,16 +261,34 @@ public unsafe BitArray And(BitArray value)
 
             int count = GetInt32ArrayLengthFromBitLength(Length);
 
+            // Unroll loop for count less than Vector256 size.
             switch (count)
             {
+                case 7: m_array[6] &= value.m_array[6]; goto case 6;
+                case 6: m_array[5] &= value.m_array[5]; goto case 5;
+                case 5: m_array[4] &= value.m_array[4]; goto case 4;
+                case 4: m_array[3] &= value.m_array[3]; goto case 3;
                 case 3: m_array[2] &= value.m_array[2]; goto case 2;
                 case 2: m_array[1] &= value.m_array[1]; goto case 1;
                 case 1: m_array[0] &= value.m_array[0]; goto Done;
                 case 0: goto Done;
             }
 
             int i = 0;
-            if (Sse2.IsSupported)
+            if (Avx2.IsSupported)
+            {
+                fixed (int* leftPtr = m_array)
+                fixed (int* rightPtr = value.m_array)
+                {
+                    for (; i < count - (Vector256<int>.Count - 1); i += Vector256<int>.Count)
+                    {
+                        Vector256<int> leftVec = Avx.LoadVector256(leftPtr + i);
```

Comment:
Are there any potential alignment issues here?

Reply:
@tannergooding Are there any Vector256 alignment issues to be aware of? (also, if you could look at this change, that would be great)

Reply:
Not from a correctness point of view as these loads/stores are unaligned anyway. It may be a perf issue as unaligned access that straddles cache lines is kind of slow. Typical vectorized implementations attempt to first align the pointer but it looks like the existing `Vector128` code didn't bother with that so...

Reply:
While I agree AVX + aligned loads + partially unrolled loop to do 4 XORs per iteration is super fast for large inputs I find SSE + unaligned loop is more friendly for small inputs and the code looks (and actually) much smaller. Just take a look at this "Calculate a sum of floats" impl with all of those in mind: https://github.com/dotnet/machinelearning/blob/287bc3e9b84f640e9b75e7b288c2663a536a9863/src/Microsoft.ML.CpuMath/AvxIntrinsics.cs#L988-L1095

Btw, the extended jump table probably regresses bitarrays with Length > 3 && Length < 8

Reply:
Yeah, as mikedn said, these are all "unaligned" loads anyways and the only real concern is for reads/writes that cross a cache-line or page boundary, which can be significantly slower. The current optimization manual indicates ~4.5 cycles vs 1 cycle for modern processors for cache-line load splits and ~150 cycles for a cache-line store split.

So, you might see some perf benefit (especially for large arrays) by aligning the pointer. You can generally still perform the first (handling unaligned elements) and last (handling trailing elements) via vectorization by masking out elements you don't care about.